components/shader/shadermanager.cpp

```diff
@@ -99,10 +99,39 @@ namespace Shader
         return true;
     }
 
-    osg::ref_ptr<osg::Shader> ShaderManager::getShader(const std::string &shaderTemplate, const ShaderManager::DefineMap &defines, osg::Shader::Type shaderType)
+    osg::ref_ptr<osg::Shader> ShaderManager::getShader(const std::string &shaderTemplate, const ShaderManager::DefineMap &defines, osg::Shader::Type shaderType, bool disableShadows)
     {
         OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
 
+        // set up shadows in the shader
+        // get these values from settings manager
+        bool shadows = true & !disableShadows;
+        int numShadowMaps = 2;
+        DefineMap definesWithShadows;
+        if (shadows)
+        {
+            definesWithShadows.insert(std::make_pair(std::string("shadows_enabled"), std::string("1")));
+
+            /*definesWithShadows.insert(std::string("shadow_texture_unit_declarations"), std::string(""));
+            definesWithShadows.insert(std::string("shadow_space_coordinate_declarations"), std::string(""));
+            definesWithShadows.insert(std::string("shadow_space_coordinate_calculations"), std::string(""));
+            definesWithShadows.insert(std::string("shadow_texture_sampler_declarations"), std::string(""));
+            definesWithShadows.insert(std::string("shadow_texture_lookup_calculations"), std::string(""));*/
+            for (int i = 0; i < numShadowMaps; ++i)
+            {
+                definesWithShadows["shadow_texture_unit_declarations"] += "uniform int shadowTextureUnit" + std::to_string(i) + ";\n";
+                definesWithShadows["shadow_space_coordinate_declarations"] += "varying vec4 shadowSpaceCoords" + std::to_string(i) + ";\n";
```

Comment:
Ignoring that this is in the wrong place (`ShaderManager` is supposed to be agnostic of specific variable names, better is 1 level up = `ShaderVisitor`), this is what I meant with generated shaders which I would really love to avoid. The line numbers are now messed up so you can no longer associate shader compile errors with their respective line, and in general the code gets much harder to grasp or to change it.

I think there's a way to declare an array of uniform values, we just might need to change the osgShadow too so it actually sets the uniform as an array. Once this is the case we could just do a for loop in the shader itself.

Reply:
I put *there* specifically because that method is called directly when things like the water shader are created. If it goes anywhere else, then it won't actually get called in all cases.

Switching to use arrays of uniforms seems like a good idea and is mostly viable... except:
* In GLSL 1.2, which is what we use, you aren't allowed arrays of samplers.
* In GLSL 1.3 to 3.3 you can't loop over arrays of samplers. (I'm guessing that it's implemented as syntactic sugar for exactly what we're doing already.)

I'm assuming we don't want to bump up the required OpenGL version to 4.x when users turn on shadows, so it may not be an option after all. Apart from that, I think I know which things I'd have to change to make `osgShadow::VDSM` upload only things we can loop over.

Reply:
Here are two possible changes that might help a bit:

* Move as much of this as possible into macros in the shaders themselves.
* Put the shadow defines in a separate map that can be accessed as a static member of `MWShadow` so it can be added manually more easily. This does, however, mean that anyone wanting to generate shaders for something needs to be aware that they have to turn on shadows themselves.

Reply:
D'oh, of course there has to be a stupid OpenGL limitation that ruins everything.

Then I'd like to put another option out there, which is to extend the shader preprocessor to support a 'for' directive that will unroll the contained code. This is actually what we did before with Ogre3D. For example:

<pre>
@foreach numshadowmaps shadowIndex
uniform sampler2D shadowTexture@shadowIndex;
@endforeach
</pre>

Would unroll to the following (assuming the '@numshadowmaps' define was set to 3):
<pre>
uniform sampler2D shadowTexture0;
uniform sampler2D shadowTexture1;
uniform sampler2D shadowTexture2;
</pre>
We'd also need to insert a #line directive after each loop make sure the line numbers stay correct.

Reply:
Wouldn't it be easier to just not put line breaks in the inserted code? GLSL doesn't care about them as it uses semicolons to separate statements.

Reply:
Do we want a `for` which iterates over numbers in a range, or a `foreach` which splits its argument into a list and iterates over those values?

Reply:
I would go with `foreach` because I've never had a usecase for `for`.

> Wouldn't it be easier to just not put line breaks in the inserted code? GLSL doesn't care about them as it uses semicolons to separate statements.

That will be problematic if the contained code actually contains line breaks. It's not very difficult to just insert the #line directive, there is an example too where we already do this when parsing `#include`s.

Reply:
I've implemented a `@foreach` preprocessor command with the input list being parsed as a comma-separated sequence going to the line end that seems to work. It's a little messier than I'd have liked, but I think that might be because of C++ and the STL not having as rich a string library as I'm used to from other languages.